hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/AtomicBatchHandler.java

```diff
@@ -0,0 +1,324 @@
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.ObjLongConsumer;
+import java.util.function.Supplier;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link HederaFunctionality#ATOMIC_BATCH}.
+ */
+@Singleton
+public class AtomicBatchHandler implements TransactionHandler {
+    private final Supplier<FeeCharging> appFeeCharging;
+
+    private static final AccountID ATOMIC_BATCH_NODE_ACCOUNT_ID =
+            AccountID.newBuilder().accountNum(0).shardNum(0).realmNum(0).build();
+    /**
+     * Constructs a {@link AtomicBatchHandler}
+     */
+    @Inject
+    public AtomicBatchHandler(@NonNull final AppContext appContext) {
+        requireNonNull(appContext);
+        this.appFeeCharging = appContext.feeChargingSupplier();
+    }
+
+    /**
+     * Performs checks independent of state or context.
+     *
+     * @param context the pure checks context
+     */
+    @Override
+    public void pureChecks(@NonNull final PureChecksContext context) throws PreCheckException {
+        requireNonNull(context);
+        final List<Transaction> innerTxs = context.body().atomicBatchOrThrow().transactions();
+        if (innerTxs.isEmpty()) {
+            throw new PreCheckException(BATCH_LIST_EMPTY);
+        }
+
+        Set<TransactionID> txIds = new HashSet<>();
+        for (final var innerTx : innerTxs) {
+            if (!innerTx.hasBody()) {
+                throw new PreCheckException(BATCH_LIST_CONTAINS_INVALID_TRANSACTION);
+            }
+            final var txBody = innerTx.bodyOrThrow(); // inner txs are required to use body
+
+            // throw if more than one tx has the same transactionID
+            validateTruePreCheck(txIds.add(txBody.transactionID()), BATCH_LIST_CONTAINS_DUPLICATES);
+
+            // validate batch key exists on each inner transaction
+            validateTruePreCheck(txBody.hasBatchKey(), MISSING_BATCH_KEY);
+
+            if (!txBody.hasNodeAccountID() || !txBody.nodeAccountIDOrThrow().equals(ATOMIC_BATCH_NODE_ACCOUNT_ID)) {
+                throw new PreCheckException(INVALID_NODE_ACCOUNT_ID);
+            }
+
+            context.dispatchPureChecks(txBody);
+        }
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var atomicBatchTransactionBody = context.body().atomicBatchOrThrow();
+        final var config = context.configuration();
+        final var atomicBatchConfig = config.getConfigData(AtomicBatchConfig.class);
+
+        for (final var innerTxBody : atomicBatchTransactionBody.transactions().stream()
```

Comment:
_Nit_ - confusing use of `stream()` when a for loop would be equally expressive.